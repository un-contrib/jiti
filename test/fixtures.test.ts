import { join, resolve, dirname } from 'pathe'
import { execa } from 'execa'
import { describe, it, expect } from 'vitest'
import fg from 'fast-glob'

describe('fixtures', async () => {
  const jitiPath = resolve(__dirname, '../bin/jiti.js')

  const root = dirname(__dirname)
  const dir = join(__dirname, 'fixtures')
  const fixtures = await fg('*/index.*', { cwd: dir })

  for (const fixture of fixtures) {
    it(dirname(fixture), async () => {
      const fixtureEntry = join(dir, fixture)
      const cwd = dirname(fixtureEntry)

      // Clean up absolute paths and sourcemap locations for stable snapshots
      function cleanUpSnap (str:string) {
        return str
          .replace(/\\/g, '/')
          .split(cwd).join('<cwd>') // workaround for replaceAll in Node 14
          .split(root).join('<root>') // workaround for replaceAll in Node 14
          .replace(/:\d+:\d+([)'\n])/g, '$1')
      }

      const { stdout, stderr } = await execa('node', [jitiPath, fixtureEntry], { cwd, stdio: 'pipe', reject: false })
      expect(cleanUpSnap(stdout)).toMatchSnapshot('stdout')
      expect(cleanUpSnap(stderr)).toMatchSnapshot('stderr')
    })
  }
})
