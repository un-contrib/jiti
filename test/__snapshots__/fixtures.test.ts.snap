// Vitest Snapshot v1

exports[`fixtures > async > stderr 1`] = `""`;

exports[`fixtures > async > stdout 1`] = `"works"`;

exports[`fixtures > circular > stderr 1`] = `""`;

exports[`fixtures > circular > stdout 1`] = `"FooBar"`;

exports[`fixtures > error-parse > stderr 1`] = `
"<root>/lib/index.js:2
  throw err /* ↓ Check stack trace ↓ */
  ^

Error: ParseError: \`import\` can only be used in \`import 
 <cwd>/index.ts
    at Object.<anonymous> (<root>/bin/jiti.js)
    at Module._compile (internal/modules/cjs/loader)
    at Object.Module._extensions..js (internal/modules/cjs/loader)
    at Module.load (internal/modules/cjs/loader)
    at Function.Module._load (internal/modules/cjs/loader)
    at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main)
    at internal/main/run_main_module:17:47"
`;

exports[`fixtures > error-parse > stdout 1`] = `""`;

exports[`fixtures > error-runtime > stderr 1`] = `
"<root>/lib/index.js:2
  throw err /* ↓ Check stack trace ↓ */
  ^

TypeError [ERR_INVALID_ARG_TYPE]: The \\"listener\\" argument must be of type function. Received undefined
    at checkListener (node:events)
    at _addListener (node:events)
    at process.addListener (node:events)
    at <cwd>/index.ts
    at g (<root>/dist/jiti.js)
    at Object.<anonymous> (<root>/bin/jiti.js)
    at Module._compile (internal/modules/cjs/loader)
    at Object.Module._extensions..js (internal/modules/cjs/loader)
    at Module.load (internal/modules/cjs/loader)
    at Function.Module._load (internal/modules/cjs/loader) {
  code: 'ERR_INVALID_ARG_TYPE'
}"
`;

exports[`fixtures > error-runtime > stdout 1`] = `""`;

exports[`fixtures > esm > stderr 1`] = `""`;

exports[`fixtures > esm > stdout 1`] = `
"{
  file: '<cwd>/test.js',
  dir: '<cwd>',
  'import.meta.url': 'file://<cwd>/test.js',
  stack: [
    'at getStack (<cwd>/test.js)',
    'at test (<cwd>/test.js)',
    'at <cwd>/index.js',
    'at g (<root>/dist/jiti.js)',
    'at Object.<anonymous> (<root>/bin/jiti.js)',
    'at Module._compile (internal/modules/cjs/loader)',
    'at Object.Module._extensions..js (internal/modules/cjs/loader)',
    'at Module.load (internal/modules/cjs/loader)',
    'at Function.Module._load (internal/modules/cjs/loader)',
    'at Function.executeUserEntryPoint [as runMain] (internal/modules/run_main)'
  ]
}"
`;

exports[`fixtures > hashbang > stderr 1`] = `""`;

exports[`fixtures > hashbang > stdout 1`] = `"1"`;

exports[`fixtures > legacy > stderr 1`] = `""`;

exports[`fixtures > legacy > stdout 1`] = `"Should be undefined: undefined"`;

exports[`fixtures > mixed > stderr 1`] = `""`;

exports[`fixtures > mixed > stdout 1`] = `"Mixed works for: <cwd>"`;

exports[`fixtures > native > stderr 1`] = `""`;

exports[`fixtures > native > stdout 1`] = `"[Module: null prototype] { default: { hasRequire: false } }"`;

exports[`fixtures > proto > stderr 1`] = `""`;

exports[`fixtures > proto > stdout 1`] = `"exists: true"`;

exports[`fixtures > typescript > stderr 1`] = `""`;

exports[`fixtures > typescript > stdout 1`] = `
"{
  file: '<cwd>/test.ts',
  dir: '<cwd>',
  resolve: '<cwd>/test.ts'
}"
`;
